search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,589 @@
+    YelpShellSearchProvider2 *skeleton;
+
+    gboolean released;
+
+    GHashTable *page_data_hash_map;
+    GPtrArray *delayed_result_getters;
+};
+
+GType yelp_search_provider_app_get_type (void);
+
+#define YELP_TYPE_SEARCH_PROVIDER_APP yelp_search_provider_app_get_type()
+#define YELP_SEARCH_PROVIDER_APP(obj) \
+      (G_TYPE_CHECK_INSTANCE_CAST((obj), YELP_TYPE_SEARCH_PROVIDER_APP, YelpSearchProviderApp))
+
+G_DEFINE_TYPE (YelpSearchProviderApp, yelp_search_provider_app, G_TYPE_APPLICATION)
+
+static PageData *
+page_data_new_steal (gchar *title,
+                     gchar *desc,
+                     GIcon *icon)
+{
+    PageData *data = g_new0 (PageData, 1);
+
+    if (title) {
+        data->title = title;
+        data->title_casefold = g_utf8_casefold (title, -1);
+    }
+    if (desc) {
+        data->desc = desc;
+        data->desc_casefold = g_utf8_casefold (desc, -1);
+    }
+    data->icon = icon;
+    return data;
+}
+
+static void
+page_data_free (PageData *page_data)
+{
+    g_free (page_data->title);
+    g_free (page_data->title_casefold);
+    g_free (page_data->desc);
+    g_free (page_data->desc_casefold);
+    g_clear_object (&page_data->icon);
+    g_free (page_data);
+}
+
+static GVariant *
+get_result_metas (gchar      **page_ids,
+                  GHashTable  *page_data)
+{
+    GVariantBuilder metas;
+    gint i;
+
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++) {
+        GVariantBuilder meta;
+        PageData *data = g_hash_table_lookup (page_data, page_ids[i]);
+        g_variant_builder_init (&meta, G_VARIANT_TYPE ("a{sv}"));
+        g_variant_builder_add (&meta, "{sv}",
+                               "id", g_variant_new_string (page_ids[i]));
+        g_variant_builder_add (&meta, "{sv}",
+                               "name", g_variant_new_string (data->title));
+        g_variant_builder_add (&meta, "{sv}",
+                               "icon", g_icon_serialize (data->icon));
+        g_variant_builder_add (&meta, "{sv}",
+                               "description", g_variant_new_string (data->desc));
+        g_variant_builder_add_value (&metas, g_variant_builder_end (&meta));
+    }
+
+    return g_variant_new ("(aa{sv})", &metas);
+}
+
+/* FIXME: Search was copied from yelp-search-entry and slightly adapted.
+ * When search in yelp gets adjusted we should export it and use it here.
+ */
+static gboolean
+match_word (const gchar *search_phrase,
+            const gchar *title_casefold,
+            const gchar *desc_casefold)
+{
+    static GRegex *nonword = NULL;
+    gint stri;
+    gchar **strs;
+    gboolean ret = TRUE;
+
+    if (nonword == NULL)
+        nonword = g_regex_new ("\\W", 0, 0, NULL);
+    g_assert (nonword != NULL);
+
+    strs = g_regex_split (nonword, search_phrase, 0);
+    for (stri = 0; strs[stri]; stri++) {
+        if (!title_casefold ||
+            !strstr (title_casefold, strs[stri]) &&
```

Comment:
I think the two parts of this `&&` condition need to be enclosed with braces. I.e.
```
if ((!a || !b) && (!c || !d)) 
```